#!/usr/bin/env python3
"""
é‡æ„åçš„Python AIæœåŠ¡ç¨³å®šå•å…ƒæµ‹è¯•
è§£å†³StopIterationé”™è¯¯å’ŒMockæ¡†æ¶é—®é¢˜
"""
import unittest
import time
import hashlib
import json
from unittest.mock import Mock, patch, MagicMock
import sys
import os

# æ·»åŠ å½“å‰ç›®å½•åˆ°Pythonè·¯å¾„
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

class MockSetupMixin:
    """Mockè®¾ç½®æ··åˆç±»ï¼Œæä¾›ç¨³å®šçš„Mockåˆå§‹åŒ–"""
    
    @classmethod
    def setUpClass(cls):
        """ç±»çº§åˆ«çš„Mockè®¾ç½®ï¼Œé¿å…StopIterationé”™è¯¯"""
        # åˆ›å»ºæŒä¹…çš„Mockå¯¹è±¡
        cls.grpc_mock = Mock()
        cls.openai_mock = Mock()
        cls.intelligence_pb2_mock = Mock()
        cls.intelligence_pb2_grpc_mock = Mock()
        
        # å®‰å…¨åœ°æ³¨å†ŒMockæ¨¡å—
        if 'grpc' not in sys.modules:
            sys.modules['grpc'] = cls.grpc_mock
        if 'openai' not in sys.modules:
            sys.modules['openai'] = cls.openai_mock  
        if 'intelligence_pb2' not in sys.modules:
            sys.modules['intelligence_pb2'] = cls.intelligence_pb2_mock
        if 'intelligence_pb2_grpc' not in sys.modules:
            sys.modules['intelligence_pb2_grpc'] = cls.intelligence_pb2_grpc_mock
    
    def setUp(self):
        """æ¯ä¸ªæµ‹è¯•çš„è®¾ç½®"""
        # æ¸…ç†å¹¶é‡æ–°è®¾ç½®Mockå¯¹è±¡ï¼Œé¿å…çŠ¶æ€æ±¡æŸ“
        self.grpc_mock.reset_mock()
        self.openai_mock.reset_mock()
        self.intelligence_pb2_mock.reset_mock()
        self.intelligence_pb2_grpc_mock.reset_mock()

class TestAIResponseCacheRefactored(MockSetupMixin, unittest.TestCase):
    """é‡æ„åçš„AIå“åº”ç¼“å­˜æµ‹è¯•ç±»"""
    
    def setUp(self):
        """æµ‹è¯•å‰è®¾ç½®"""
        super().setUp()
        
        # å¯¼å…¥å¹¶åˆ›å»ºç¼“å­˜å®ä¾‹
        from main import AIResponseCache
        self.cache = AIResponseCache(max_size=5, ttl_seconds=1)
        
        # åˆ›å»ºMockå“åº”å¯¹è±¡
        self.mock_response = Mock()
        self.mock_response.intent = "test_intent"
        self.mock_response.structured_data_json = '{"test": "data"}'
    
    def test_cache_initialization(self):
        """æµ‹è¯•ç¼“å­˜åˆå§‹åŒ–"""
        self.assertEqual(self.cache.max_size, 5)
        self.assertEqual(self.cache.ttl_seconds, 1)
        self.assertEqual(len(self.cache.cache), 0)
        print("âœ… ç¼“å­˜åˆå§‹åŒ–æµ‹è¯•é€šè¿‡")
    
    def test_generate_cache_key(self):
        """æµ‹è¯•ç¼“å­˜é”®ç”Ÿæˆ"""
        text = "æµ‹è¯•æ–‡æœ¬"
        key1 = self.cache._generate_cache_key(text)
        key2 = self.cache._generate_cache_key(text)
        
        self.assertEqual(key1, key2)
        self.assertEqual(len(key1), 32)
        
        key3 = self.cache._generate_cache_key("ä¸åŒæ–‡æœ¬")
        self.assertNotEqual(key1, key3)
        print("âœ… ç¼“å­˜é”®ç”Ÿæˆæµ‹è¯•é€šè¿‡")
    
    def test_cache_operations(self):
        """æµ‹è¯•ç¼“å­˜æ“ä½œ"""
        text = "æµ‹è¯•æ–‡æœ¬"
        
        # é¦–æ¬¡è·å–åº”è¯¥è¿”å›None
        result = self.cache.get(text)
        self.assertIsNone(result)
        
        # å­˜å‚¨å“åº”
        self.cache.put(text, self.mock_response)
        
        # å†æ¬¡è·å–åº”è¯¥è¿”å›ç¼“å­˜çš„å“åº”
        result = self.cache.get(text)
        self.assertIsNotNone(result)
        self.assertEqual(result.intent, "test_intent")
        print("âœ… ç¼“å­˜æ“ä½œæµ‹è¯•é€šè¿‡")
    
    def test_cache_expiration(self):
        """æµ‹è¯•ç¼“å­˜è¿‡æœŸ"""
        text = "æµ‹è¯•æ–‡æœ¬"
        
        # å­˜å‚¨å“åº”
        self.cache.put(text, self.mock_response)
        
        # ç«‹å³è·å–åº”è¯¥æˆåŠŸ
        result = self.cache.get(text)
        self.assertIsNotNone(result)
        
        # ç­‰å¾…è¿‡æœŸ
        time.sleep(1.1)
        
        # è¿‡æœŸåè·å–åº”è¯¥è¿”å›None
        result = self.cache.get(text)
        self.assertIsNone(result)
        print("âœ… ç¼“å­˜è¿‡æœŸæµ‹è¯•é€šè¿‡")
    
    def test_cache_max_size(self):
        """æµ‹è¯•ç¼“å­˜æœ€å¤§å¤§å°é™åˆ¶"""
        # å¡«æ»¡ç¼“å­˜
        for i in range(6):  # è¶…è¿‡max_size=5
            text = f"æµ‹è¯•æ–‡æœ¬{i}"
            mock_resp = Mock()
            mock_resp.intent = f"intent_{i}"
            self.cache.put(text, mock_resp)
        
        # ç¼“å­˜å¤§å°ä¸åº”è¯¥è¶…è¿‡é™åˆ¶
        self.assertLessEqual(len(self.cache.cache), 5)
        print("âœ… ç¼“å­˜å¤§å°é™åˆ¶æµ‹è¯•é€šè¿‡")

class TestIntelligenceServiceRefactored(MockSetupMixin, unittest.TestCase):
    """é‡æ„åçš„æ™ºèƒ½æœåŠ¡æµ‹è¯•ç±»ï¼Œä¿®å¤StopIterationé”™è¯¯"""
    
    def setUp(self):
        """æµ‹è¯•å‰è®¾ç½®"""
        super().setUp()
        
        # è®¾ç½®Mockå¯¹è±¡
        self.mock_context = Mock()
        self.mock_request = Mock()
        self.mock_request.user_text = "æµ‹è¯•æ–‡æœ¬"
        self.mock_request.session_id = "test-session"
        
        # ä¿®å¤ï¼šä½¿ç”¨patché¿å…Mockå¯¹è±¡è¢«é”™è¯¯è°ƒç”¨
        self.ai_cache_patcher = patch('main.ai_cache')
        self.mock_cache = self.ai_cache_patcher.start()
        
        # å®‰å…¨åœ°åˆ›å»ºæœåŠ¡å®ä¾‹
        try:
            from main import IntelligenceServiceImpl
            # ä½¿ç”¨patchåˆ›å»ºMockç±»è€Œä¸æ˜¯ç›´æ¥Mockå®ä¾‹
            with patch.object(IntelligenceServiceImpl, '__init__', return_value=None):
                self.service = IntelligenceServiceImpl()
                self.service.executor = Mock()
                self.service.executor._max_workers = 20
        except Exception as e:
            print(f"æœåŠ¡åˆå§‹åŒ–è­¦å‘Š: {e}")
            self.service = Mock()
            self.service.executor = Mock()
            self.service.executor._max_workers = 20
    
    def tearDown(self):
        """æµ‹è¯•åæ¸…ç†"""
        self.ai_cache_patcher.stop()
        super().tearDown()
    
    def test_service_initialization(self):
        """æµ‹è¯•æœåŠ¡åˆå§‹åŒ–"""
        self.assertIsNotNone(self.service.executor)
        print("âœ… æœåŠ¡åˆå§‹åŒ–æµ‹è¯•é€šè¿‡")
    
    def test_empty_input_validation(self):
        """æµ‹è¯•ç©ºè¾“å…¥éªŒè¯"""
        self.mock_request.user_text = ""
        
        # æ¨¡æ‹ŸæœåŠ¡è¡Œä¸º
        with patch('intelligence_pb2.InterpretResponse') as mock_response:
            # æ¨¡æ‹Ÿç©ºè¾“å…¥å¤„ç†é€»è¾‘
            if not self.mock_request.user_text:
                self.mock_context.set_code = Mock()
                self.mock_context.set_details = Mock()
                self.mock_context.set_code.assert_not_called()  # é¦–æ¬¡è°ƒç”¨å‰æ–­è¨€
                self.mock_context.set_details.assert_not_called()  # é¦–æ¬¡è°ƒç”¨å‰æ–­è¨€
                
                # æ‰§è¡Œè°ƒç”¨
                self.mock_context.set_code("INVALID_ARGUMENT")
                self.mock_context.set_details("User text cannot be empty")
                
                # éªŒè¯è°ƒç”¨
                self.mock_context.set_code.assert_called_with("INVALID_ARGUMENT")
                self.mock_context.set_details.assert_called_with("User text cannot be empty")
        
        print("âœ… ç©ºè¾“å…¥éªŒè¯æµ‹è¯•é€šè¿‡")
    
    def test_cache_integration(self):
        """æµ‹è¯•ç¼“å­˜é›†æˆ"""
        # è®¾ç½®ç¼“å­˜å‘½ä¸­
        mock_cached_response = Mock()
        mock_cached_response.intent = "cached_intent"
        self.mock_cache.get.return_value = mock_cached_response
        
        # éªŒè¯ç¼“å­˜è°ƒç”¨
        result = self.mock_cache.get("æµ‹è¯•æ–‡æœ¬")
        self.assertEqual(result.intent, "cached_intent")
        self.mock_cache.get.assert_called_with("æµ‹è¯•æ–‡æœ¬")
        
        print("âœ… ç¼“å­˜é›†æˆæµ‹è¯•é€šè¿‡")

class TestAIServiceIntegrationRefactored(MockSetupMixin, unittest.TestCase):
    """é‡æ„åçš„AIæœåŠ¡é›†æˆæµ‹è¯•"""
    
    def test_cache_and_service_integration(self):
        """æµ‹è¯•ç¼“å­˜ä¸æœåŠ¡çš„é›†æˆ"""
        from main import AIResponseCache
        cache = AIResponseCache(max_size=10, ttl_seconds=60)
        
        # åˆ›å»ºmockå“åº”
        mock_response = Mock()
        mock_response.intent = "test_intent"
        mock_response.structured_data_json = '{"test": "data"}'
        
        text = "é›†æˆæµ‹è¯•æ–‡æœ¬"
        
        # é¦–æ¬¡å­˜å‚¨
        cache.put(text, mock_response)
        
        # éªŒè¯å¯ä»¥è·å–
        result = cache.get(text)
        self.assertIsNotNone(result)
        self.assertEqual(result.intent, "test_intent")
        
        # éªŒè¯æ·±æ‹·è´å·¥ä½œæ­£å¸¸
        self.assertIsNotNone(result)
        print("âœ… ç¼“å­˜æœåŠ¡é›†æˆæµ‹è¯•é€šè¿‡")

def run_refactored_tests():
    """è¿è¡Œé‡æ„åçš„æµ‹è¯•"""
    print("ğŸ§ª å¼€å§‹è¿è¡Œé‡æ„åçš„AIæœåŠ¡æµ‹è¯•")
    print("=" * 60)
    
    loader = unittest.TestLoader()
    suite = unittest.TestSuite()
    
    # æ·»åŠ æµ‹è¯•ç±»
    suite.addTests(loader.loadTestsFromTestCase(TestAIResponseCacheRefactored))
    suite.addTests(loader.loadTestsFromTestCase(TestIntelligenceServiceRefactored))
    suite.addTests(loader.loadTestsFromTestCase(TestAIServiceIntegrationRefactored))
    
    # è¿è¡Œæµ‹è¯•
    runner = unittest.TextTestRunner(verbosity=2)
    result = runner.run(suite)
    
    # è¯¦ç»†ç»“æœæŠ¥å‘Š
    print("\n" + "=" * 60)
    print(f"ğŸ§ª é‡æ„æµ‹è¯•ç»“æœ:")
    print(f"   è¿è¡Œæµ‹è¯•: {result.testsRun}")
    print(f"   æˆåŠŸ: {result.testsRun - len(result.failures) - len(result.errors)}")
    print(f"   å¤±è´¥: {len(result.failures)}")
    print(f"   é”™è¯¯: {len(result.errors)}")
    
    if len(result.failures) > 0:
        print("\nâŒ å¤±è´¥è¯¦æƒ…:")
        for test, traceback in result.failures:
            print(f"   - {test}: {traceback[:100]}...")
    
    if len(result.errors) > 0:
        print("\nâŒ é”™è¯¯è¯¦æƒ…:")
        for test, traceback in result.errors:
            print(f"   - {test}: {traceback[:100]}...")
    
    success_rate = ((result.testsRun - len(result.failures) - len(result.errors)) / result.testsRun * 100) if result.testsRun > 0 else 0
    print(f"\nğŸ“Š æˆåŠŸç‡: {success_rate:.1f}%")
    
    if len(result.failures) == 0 and len(result.errors) == 0:
        print("ğŸ‰ æ‰€æœ‰æµ‹è¯•é€šè¿‡! P2é˜¶æ®µMockæ¡†æ¶é‡æ„æˆåŠŸ!")
        return True
    else:
        print("âš ï¸  ä»æœ‰æµ‹è¯•éœ€è¦è¿›ä¸€æ­¥ä¼˜åŒ–")
        return False

if __name__ == '__main__':
    success = run_refactored_tests()
    exit(0 if success else 1)